### ISSUE TITLE
train-dataset

### ISSUE THREAD

**Issue by zlm-ai** [Time-MoE/Time-MoE#15]
Your work is of great significance. I would like to reproduce your small - scale model. However, the time_moe_dataset class in git seems to be designed for evaluation. I wonder if you can share the training dataset code.

**Comment by Maple728** [comment]
Thank you for your interest in our work.

The time_moe_dataset is indeed used in the training dataset; however, it needs to be encapsulated within a dataset that converts a time series into multiple fixed-window-size time series, which can then be organized into batch data.

We are currently preparing the code for fine-tuning forecasting, and it will be released in the coming week or two.

Please feel free to reach out if you have any further questions.

**Comment by zlm-ai** [comment]
Thank you for your reply. I got it. I'm really looking forward to your fine - tuning code.

**Comment by zlm-ai** [comment]
On the other hand, I encountered some out - of - memory issues when instantiating TimeMoeForPrediction. I instantiated the model on an A100 graphics card with 40G of video memory. My instantiation code is as follows:
`config = TimeMoeConfig()
model = TimeMoeForPrediction(config) #.to(device)`

**Comment by Maple728** [comment]
Could you please provide the complete code statement?

P.S. We recommend using flash-attention if you have an A100 or H100 GPU, as it will be faster and more memory-efficient.

**Comment by zlm-ai** [comment]
My code is as follows.
`device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

config = TimeMoeConfig()
model = TimeMoeForPrediction(config) #.to(device)

criterion = nn.HuberLoss(reduction='mean', delta=2.0)
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

train_dataset = time_moe_train_dataset.TimeMoEDataset(data_folder,batch_size)

for epoch in range(num_epochs):
    model.train()
    total_loss = 0.0
    for batch, attn_mask in tqdm(train_dataset, desc=f'Epoch {epoch + 1}/{num_epochs}'):
        # batch = batch.to(device)
        # attn_mask = attn_mask.to(device)
        outputs = model.forward(input_ids=batch,
                                attention_mask=attn_mask,
                                return_dict=True)
        loss = outputs.loss if isinstance(outputs, MoeCausalLMOutputWithPast) else outputs[0]

        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        total_loss += loss.item()
 I'm trying to construct an instance model of TimeMoeForPrediction. An out - of - memory situation occurs when executing
`model = TimeMoeForPrediction(config) .to(device)`

**Comment by Maple728** [comment]
What is the length of the sequence generated by `train_dataset`? If the sequence is very long, memory consumption can become quite high when using the `eager` attention function instead of `flash-attention`.

**Comment by zlm-ai** [comment]
> What is the length of the sequence generated by `train_dataset`? If the sequence is very long, memory consumption can become quite high when using the `eager` attention function instead of `flash-attention`.

The train_dataset is the sale part of the 300 - billion - scale dataset. But in fact, we encountered an out-of-memory situation when instantiating the 32 - layer TimeMoeDecoderLayer.

**Comment by Maple728** [comment]
The fine-tuning code is now available! You can find the tutorial in the `README.md`. I hope you find it helpful.

If you have any further questions, please don't hesitate to contact us.

**Comment by zlm-ai** [comment]
> > The train_dataset is the sale part of the 300 - billion - scale dataset. But in fact, we encountered an out-of-memory situation when instantiating the 32 - layer TimeMoeDecoderLayer.
> 
> The Time-300B dataset features time-series with very long sequences, which can result in significant memory consumption if fed directly into the model. A common approach to address this is to split the original time-series into multiple fixed-length sequences before inputting them into the model.
> 
> The fine-tuning code is now available! You can find the tutorial in the `README.md`. I hope you find it helpful.
> 
> If you have any further questions, please don't hesitate to contact us.


Thank you very much for your reply. Your help is of great significance to me.
